0-bootstrap/modules/jenkins/files/jenkins_gce_startup_script.sh

```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#!/bin/sh
+
+echo "**** Startup Step 1/9: Update apt-get repositories. ****"
+sudo apt-get update
+
+echo "**** Startup Step 2/8: Install Java. ****"
+sudo apt-get install -y default-jdk
+
+echo "**** Startup Step 3/8: Create the directory for Terraform in /opt. ****"
```

Comment:
if we are installing the dependencies locally, we should align with the bootstrap / dependencies installed in the container in the boostrap module https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/master/modules/cloudbuild/cloudbuild_builder/Dockerfile

Reply:
The broader question is, if we are using containers for builds why do we need these dependencies on the VM?

Reply:
There are still dependencies in that docker file which we should include such as jq & alpha / beta components for gcloud. It would be nice to also verify the shasum of the terraform binary given you are passing that information in

Reply:
I can just implement the same apt-get we have in the Dockerfile:
```
apt-get update && \
   /builder/google-cloud-sdk/bin/gcloud -q components install alpha beta && \
    apt-get -y install curl jq unzip ca-certificates && \
    curl https://releases.hashicorp.com/terraform/${ENV_TERRAFORM_VERSION}/terraform_${ENV_TERRAFORM_VERSION}_linux_amd64.zip \
      > terraform_linux_amd64.zip && \
    echo "${ENV_TERRAFORM_VERSION_SHA256SUM} terraform_linux_amd64.zip" > terraform_SHA256SUMS && \
    sha256sum -c terraform_SHA256SUMS --status && \
    unzip terraform_linux_amd64.zip -d /builder/terraform && \
    rm -f terraform_linux_amd64.zip && \
    apt-get remove --purge -y curl unzip && \
    apt-get --purge -y autoremove && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
```
But, we also have the option of just using the Dockerfile for the pipeline... I am thinking about this because I know you are also working in a standard way of running the pipeline (using the bash scripts).

So, is the greater good to use the Dockerfile? or to install all these binaries in the GCE instance? - If we use the Dockerfile, then we just need to install docker in the GCE instance.

Please let me know what do you think and how to proceed.

Cheers.

Reply:
@rjerrems and @bharathkkb @morgante -
I just discovered an issue with the installation on packages: The Agent cannot download any package if it doesn't have an external IP. We are not adding an external IP in the final version of the code, because all communication will happen with on-prem and thru the VPN.

So, the solution is to use the container optimised OS (instead of Debian) and use docker in the pipeline. This way we don't have to install Docker either.

The only problem I see is... How will the docker image be built without internet access? - Using private repositories?

Reply:
Ultimately the agent is going to need internet access in some form, potentially through Cloud NAT. Even if it can reach the APIs privately, it is still going to need to reach Github / Terraform registry to pull the CFT modules as a part of the build process so we will need to solve this problem, unless spend time investigating options to cache the modules someone that is reachable without internet access

Reply:
Thank you @rjerrems!

The team agreed on using Cloud NAT for now and I've implemented that as the solution for this issue. We have in backlog to do something more sophisticated, such as getting binaries and packages from on-prem (thru the VPN) or to filter the websites the Agent will be able to reach thru the NAT.

Reply:
No worries -  I have unresolved this one, there are still a couple of things missing that are mentioned above. Most importantly:

- jq (used in a number of the modules)
- beta / alpha commands for gcloud

Reply:
Yes, doing that right now actually. Will send a push later todays with it.